src/utils/reader.py

```diff
@@ -1,3 +1,223 @@
+import io
+import os
+import struct
+from typing import Optional, Union
+
+import click
+
+UINT64_SIZE_IN_BYTES = 8
+UINT32_SIZE_IN_BYTES = 4
+CHAR_SIZE_IN_BYTES = 1
+DOUBLE_SIZE_IN_BYTES = 8
+FLOAT_SIZE_IN_BYTES = 4
+
+NUM_BYTES_TIMESTAMP = UINT64_SIZE_IN_BYTES
+NUM_BYTES_TRANSLATION = 3 * DOUBLE_SIZE_IN_BYTES
+NUM_BYTES_ROTATION = 4 * DOUBLE_SIZE_IN_BYTES
+NUM_BYTES_DIMENSION = UINT32_SIZE_IN_BYTES
+NUM_BYTES_FEELINGS = 4 * FLOAT_SIZE_IN_BYTES
+NUM_BYTES_PIXEL_COLOR_IMAGE = 3  # RGB format.
+NUM_BYTES_PIXEL_DEPTH_IMAGE = FLOAT_SIZE_IN_BYTES
+CONSTANT_NUM_OF_BYTES_IN_SNAPSHOT = NUM_BYTES_TIMESTAMP + \
+    NUM_BYTES_TRANSLATION + \
+    NUM_BYTES_ROTATION + \
+    (4 * NUM_BYTES_DIMENSION) + \
+    NUM_BYTES_FEELINGS
+
+
 class Reader:
-    # TODO: implement this class as part of exercise 6.
-    pass
+    def __init__(self, path: str) -> None:
+        self.file: io.BufferedReader = open(path, "rb")
+        self.id: int
+        self.name: str
+        self.birth_day: int
+        self.gender: str
+        self.read_sample_metadata()
+
+        print(f"{self.id=}, {self.name=}, {self.birth_day=}, {self.gender=}")
+
+    def __iter__(self):
+        while True:
+            curr_snapshot = Snapshot.read_from_file(file=self.file)
+            if curr_snapshot is None:
+                break
+            yield curr_snapshot
+
+    @property
+    def user_id(self) -> int:
+        return self.id
+
+    @property
+    def username(self) -> str:
+        return self.name
+
+    def read_sample_metadata(self):
+        self.id: int = from_bytes(
+            data=self.file.read(UINT64_SIZE_IN_BYTES),
+            data_type="uint64",
+            endianness="<",
+        )
+        name_length: int = from_bytes(
+            data=self.file.read(UINT32_SIZE_IN_BYTES),
+            data_type="uint32",
+            endianness="<",
+        )
+        self.name: str = from_bytes(
+            data=self.file.read(name_length),
+            data_type="string",
+            endianness="<",
+            num_bytes=name_length,
+        )
+        self.birth_day: int = from_bytes(
+            data=self.file.read(UINT32_SIZE_IN_BYTES),
+            data_type="uint32",
+            endianness="<",
+        )
+        self.gender: str = from_bytes(
+            data=self.file.read(CHAR_SIZE_IN_BYTES),
+            data_type="char",
+            endianness="<",
+        )
+
+
+class Snapshot:
```

Comment:
Please add documentation for the snapshot class.

Reply:
It will be done in a different PR.